lib/LDA_CTR.py

```diff
@@ -0,0 +1,123 @@
+            self.document_distribution = matrix
+            if self._verbose and matrix_found:
+                print("Document distribution was set from file, will not train.")
+        if matrix_found is False:
+            if self._verbose and self._load_matrices:
+                print("Document distribution file was not found, will train LDA.")
+            self._train()
+            if self._dump_matrices:
+                self.initializer.save_matrix(self.document_distribution, 'document_distribution_lda')
+        self.user_vecs = numpy.random.random((self.n_users, self.n_factors))
+        self.training_data_mask = self.evaluator.get_fold_training_data_mask()
+        confidence_matrix = numpy.zeros(self.train_data.shape)
+        confidence_matrix[self.training_data_mask] = self.train_data[self.training_data_mask]
+        zeros_mask = self.train_data == 0
+        training_zeros_mask = numpy.logical_and(self.training_data_mask, zeros_mask)
+        confidence_matrix[training_zeros_mask] = 0.01
+        self.user_vecs = self.als_step(self.user_vecs, self.document_distribution, self.train_data, self._lambda, confidence_matrix)
+        if return_report:
+            return self.get_evaluation_report()
+
+    def _train(self):
+        """
+        Train LDA Recommender, and store the document_distribution.
+        """
+        term_freq = self.abstracts_preprocessor.get_term_frequency_sparse_matrix()
+        lda = LatentDirichletAllocation(n_topics=self.n_factors, max_iter=self.n_iter,
+                                        learning_method='online',
+                                        learning_offset=50., random_state=0,
+                                        verbose=0)
+        if self._verbose:
+            print("Initialized LDA model..., Training LDA...")
+
+        self.document_distribution = lda.fit_transform(term_freq)
+        if self._verbose:
+            print("LDA trained..")
+    
+
+    def als_step(self, latent_vectors, fixed_vecs, ratings, _lambda, confidence_matrix):
+        """
+        The function computes only one step in the ALS algorithm
+
+        :param ndarray latent_vectors: the vector to be optimized
+        :param ndarray fixed_vecs: the vector to be fixed
+        :param ndarray ratings: ratings that will be used to optimize latent * fixed
+        :param float _lambda: reguralization parameter
+        """
+        numpy.where(">1")
+        lambdaI = numpy.eye(self.n_factors) * _lambda
+        for u in range(latent_vectors.shape[0]):
+            """
+            Omar code:
+            confidence = self.build_confidence_matrix(u, 'user')
+            YTY = (fixed_vecs.T * confidence).dot(fixed_vecs)
+            latent_vectors[u, :] = solve((YTY + lambdaI), (ratings[u, :] * confidence).dot(fixed_vecs))
+            """
+            # Anas Code
+            YTY = (fixed_vecs.T * confidence_matrix[u]).dot(fixed_vecs)
```

Comment:
What kind of update that we achieve by this loop? I don't see latent_vectors updated.